crates/bevy_ecs/src/component/mod.rs

```diff
@@ -86,222 +116,141 @@ impl ComponentInfo {
     pub fn is_send_and_sync(&self) -> bool {
         self.is_send_and_sync
     }
+}
 
-    fn new(id: ComponentId, descriptor: ComponentDescriptor) -> Self {
-        ComponentInfo {
-            id,
-            name: descriptor.name,
-            storage_type: descriptor.storage_type,
-            type_id: descriptor.type_id,
-            is_send_and_sync: descriptor.is_send_and_sync,
-            drop: descriptor.drop,
-            layout: descriptor.layout,
+impl From<TypeInfo> for ComponentDescriptor {
+    fn from(type_info: TypeInfo) -> Self {
+        Self {
+            name: type_info.type_name().to_string(),
+            storage_type: StorageType::default(),
+            is_send_and_sync: type_info.is_send_and_sync(),
+            type_id: Some(type_info.type_id()),
+            drop: type_info.drop(),
+            layout: type_info.layout(),
         }
     }
 }
 
-#[derive(Debug, Copy, Clone, Hash, Ord, PartialOrd, Eq, PartialEq)]
-pub struct ComponentId(usize);
-
-impl ComponentId {
-    #[inline]
-    pub const fn new(index: usize) -> ComponentId {
-        ComponentId(index)
-    }
+#[derive(Debug, Copy, Clone, Hash, Eq, PartialEq, PartialOrd, Ord)]
+pub struct RelationKindId(usize);
 
-    #[inline]
-    pub fn index(self) -> usize {
-        self.0
-    }
-}
-
-impl SparseSetIndex for ComponentId {
+impl SparseSetIndex for RelationKindId {
     #[inline]
     fn sparse_set_index(&self) -> usize {
-        self.index()
+        self.0
     }
 
     fn get_sparse_set_index(value: usize) -> Self {
         Self(value)
     }
 }
 
-pub struct ComponentDescriptor {
-    name: String,
-    storage_type: StorageType,
-    // SAFETY: This must remain private. It must only be set to "true" if this component is
-    // actually Send + Sync
-    is_send_and_sync: bool,
-    type_id: Option<TypeId>,
-    layout: Layout,
-    drop: unsafe fn(*mut u8),
+#[derive(Debug)]
+pub struct RelationKindInfo {
+    data: ComponentDescriptor,
+    id: RelationKindId,
 }
 
-impl ComponentDescriptor {
-    pub fn new<T: Component>(storage_type: StorageType) -> Self {
-        Self {
-            name: std::any::type_name::<T>().to_string(),
-            storage_type,
-            is_send_and_sync: true,
-            type_id: Some(TypeId::of::<T>()),
-            layout: Layout::new::<T>(),
-            drop: TypeInfo::drop_ptr::<T>,
-        }
-    }
-
-    #[inline]
-    pub fn storage_type(&self) -> StorageType {
-        self.storage_type
-    }
-
-    #[inline]
-    pub fn type_id(&self) -> Option<TypeId> {
-        self.type_id
+impl RelationKindInfo {
+    pub fn data_layout(&self) -> &ComponentDescriptor {
```

Comment:
```suggestion
    #[inline]
    pub fn data_layout(&self) -> &ComponentDescriptor {
```
A call to this function has higher overhead than the actual body of this function.